tldw_Server_API/app/core/Research/discovery/catalog.py

```
@@ -0,0 +1,261 @@
+
+    def resolve_selection(
+        self,
+        source_ids: list[str],
+        categories: list[str],
+    ) -> tuple[list[ResearchSourceCatalogEntry], SourceSelectionError | None]:
+        """Expand source/category selections, dedupe, and enforce the source cap."""
+        unknown_source_ids = [source_id for source_id in source_ids if source_id not in self._entries_by_id]
+        if unknown_source_ids:
+            return [], SourceSelectionError(
+                code="unknown_source",
+                message=("Unknown research discovery source id(s): " f"{', '.join(unknown_source_ids)}."),
```

Ruff (0.15.17)

[warning] Implicitly concatenated string literals on one line

(ISC001)